snapcraft_legacy/plugins/v2/catkin.py

```diff
@@ -86,26 +98,61 @@ def _get_workspace_activation_commands(self) -> List[str]:
         specific functionality.
         """
 
-        # There are a number of unbound vars, disable flag
-        # after saving current state to restore after.
-        return [
-            'state="$(set +o); set -$-"',
-            "set +u",
-            'if [ -f "${SNAPCRAFT_PART_INSTALL}/opt/ros/${ROS_DISTRO}/setup.sh" ]; then',
-            "set -- --local",
-            '_CATKIN_SETUP_DIR="{path}" . "{path}/setup.sh"'.format(
-                path="${SNAPCRAFT_PART_INSTALL}/opt/ros/${ROS_DISTRO}"
-            ),
-            "set -- --local --extend",
-            "else",
-            "set -- --local",
-            "fi",
-            '. /opt/ros/"${ROS_DISTRO}"/setup.sh',
-            'eval "${state}"',
-        ]
+        activation_commands = list()
+
+        # Source ROS ws in all build-snaps first
+        activation_commands.append("## Sourcing ROS ws in build snaps")
+        if self.options.build_snaps:
+            for build_snap in self.options.build_snaps:
+                snap_name = _get_parsed_snap(build_snap)[0]
+                activation_commands.extend(self._get_source_command(f"/snap/{snap_name}/current"))
+            activation_commands.append("")
+
+        # Source ROS ws in stage-snaps next
+        activation_commands.append("## Sourcing ROS ws in stage snaps")
+        activation_commands.extend(self._get_source_command("${SNAPCRAFT_PART_INSTALL}"))
+        activation_commands.append("")
+
+        # Finally source system's ROS ws
+        activation_commands.append("## Sourcing ROS ws in system")
+        activation_commands.extend(self._get_source_command(""))
+        activation_commands.append("")
+
+        return activation_commands
 
     def _get_build_commands(self) -> List[str]:
-        cmd = [
+
+        prepare_build_command = list()
+        export_command = list()
+
+        if self.options.build_snaps:
+            for build_snap in self.options.build_snaps:
+                build_snap = _get_parsed_snap(build_snap)[0]
+                # ROS workspaces present in buid-snaps have their '*.cmake' files pointing to hard-coded internal paths.
+                # We need to fix those paths so that they can be properly sourced.
+                # Since snaps are immutable, we copy the said '*.cmake' files to the current SNAPCRAFT_PART_BUILD path
+                # and edit these copies accordingly.
```

Comment:
Shouldn't this be done once and for all within the content-sharing snap when we build them?

Reply:
We could do that overriding the build step but that would prevent the use of custom snaps. Users would have to do that too for their own snap.